api/v1alpha1/gateway_types.go

```diff
@@ -180,6 +188,66 @@ type ListenerTLS struct {
 	Options map[string]string `json:"options"`
 }
 
+// TLSTerminationPolicy defines the schema of a TLS termination policy.
+type TLSTerminationPolicy struct {
+	// TerminationType defines how to terminate TLS connections.
+	//
+	// If unspecified, TLS termination type "Edge" will be used.
+	//
+	// Support: Core
+	//
+	// +optional
+	TerminationType TLSTerminationType `json:"terminationType,omitempty"`
+	// Routes is one or more routes to associate with the TLS termination policy.
+	//
+	// If unspecified, all routes associated with the Gateway listener will be used.
+	//
+	// Support: Core
+	//
+	// +optional
+	Routes []core.ObjectReference `json:"routes,omitempty"`
+	// CACertificates is a reference to one or more CA certificate used
+	// for establishing a TLS connection with the final destination when
+	// using TLS termination type "Reencrypt".
+	//
+	// Here is a ConfigMap example (in yaml):
+	//
+	// apiVersion: v1
+	// kind: ConfigMap
+	// metadata:
+	//  name: my-dest-svc-ca
+	//  namespace: my-dest-svc-namespace
+	//  data:
+	//    ca-bundle.crt: |
+	//      -----BEGIN CERTIFICATE-----
+	//      Destination Service CA Certificate Bundle.
+	//      -----END CERTIFICATE-----
+	//
+	// Support: Core (Kubernetes ConfigMap)
+	// Support: Implementation-specific (For other resource types)
+	//
+	// +optional
+	CACertificates []core.TypedLocalObjectReference `json:"caCertificates,omitempty"`
+}
+
+// TLSTerminationType specifies where TLS connections will terminate.
+type TLSTerminationType string
+
+const (
+	// TLSTerminationEdge terminates the TLS connection at the gateway.
+	TLSTerminationEdge TLSTerminationType = "Edge"
+
+	// TLSTerminationPassthrough terminates the TLS connection at the
+	// destination service. The destination service is responsible for
+	// decrypting data from the connection.
+	TLSTerminationPassthrough TLSTerminationType = "Passthrough"
+
+	// TLSTerminationReencrypt terminates the TLS connection at the gateway.
+	// The gateway creates an encrypted connection to the destination service
+	// using the provided certificate from DestinationCACertificate.
+	TLSTerminationReencrypt TLSTerminationType = "Reencrypt"
```

Comment:
I'm not really sure that there are 3 cases. Another way to look at it is that we have 2 independent decisions
- terminate TLS
  - proxy over TCP
  - proxy over TLS
- don't terminate TLS
  - proxy over TCP
  - proxy over TLS

So, in this view, re-encryption is a property of the backend route.

Reply:
I see where you're coming from. "Reencrypt" could be a route filter.